modules/tensor_mechanics/src/actions/ProjectedStatefulMaterialStorageAction.C

```diff
@@ -0,0 +1,235 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "ProjectedStatefulMaterialStorageAction.h"
+#include "RankFourTensorForward.h"
+#include "RankTwoTensorForward.h"
+#include "AddAuxVariableAction.h"
+#include "AddVariableAction.h"
+#include "Conversion.h"
+#include "FEProblemBase.h"
+#include "Factory.h"
+
+#include "libmesh/string_to_enum.h"
+
+// we need to add these variables after the material property types are known
+registerMooseAction("TensorMechanicsApp",
+                    ProjectedStatefulMaterialStorageAction,
+                    "setup_projected_properties");
+
+registerMooseAction("MooseApp", ProjectedStatefulMaterialStorageAction, "add_aux_kernel");
+
+InputParameters
+ProjectedStatefulMaterialStorageAction::validParams()
+{
+  InputParameters params = Action::validParams();
+
+  params.addClassDescription("Mark material properties for projected stateful storage.");
+
+  params.addParam<MooseEnum>(
+      "family",
+      MooseEnum("LAGRANGE MONOMIAL", "LAGRANGE"),
```

Comment:
I don't see a need to restrict these. Might as well allow all the families

Reply:
yeah? How would this work for HERMITE? Nodal patch recovery does not support this type of shape function as far as I can see.

Reply:
In my perfect world we would use @roystgnr's `GenericProjector`. And in that case we would add the requisite `eval_at_point` and `eval_at_node` APIs to your nodal patch recovery user object. But this would take some time to implement. For instance in the class that is used as the `FFunctor` in `system_projection.C`, `eval_at_point` is trivial to implement ... you just reinit an FE object at the requested `Point`. However, our material properties are already evaluated arrays. If one gives a `Material` a `Point` and says "what is the material property evaluation there?", we would prefer to return from our already evaluated arrays which are indexed with integers